internal/pkg/source/source.go

```diff
@@ -119,73 +119,59 @@ func sourcesFromBlocks(fn *ssa.Function, conf *config.Config, taggedFields field
 	var sources []*Source
 	for _, b := range fn.Blocks {
 		for _, instr := range b.Instrs {
-			// This type switch is used to catch instructions that could produce sources.
-			// All instructions that do not match one of the cases will hit the "default"
-			// and they will not be examined any further.
-			switch v := instr.(type) {
-			// drop anything that doesn't match one of the following cases
-			default:
-				continue
-
-			// Values produced by sanitizers are not sources.
-			case *ssa.Alloc:
-				if isProducedBySanitizer(v, conf) {
-					continue
-				}
-
-			// Values produced by sanitizers are not sources.
-			// Values produced by field propagators are.
-			case *ssa.Call:
-				if isProducedBySanitizer(v, conf) {
-					continue
-				}
-
-				if propagators.IsFieldPropagator(v) {
-					sources = append(sources, New(v))
-					continue
-				}
-
-			// A panicky type assert like s := e.(*core.Source) does not result in ssa.Extract
-			// so we need to create a source if the type assert is panicky i.e. CommaOk is false
-			// and the type being asserted is a source type.
-			case *ssa.TypeAssert:
-				if !v.CommaOk && IsSourceType(conf, taggedFields, v.AssertedType) {
-					sources = append(sources, New(v))
-				}
-
-			// An Extract is used to obtain a value from an instruction that returns multiple values.
-			// If the extracted value is a Pointer to a Source, it won't have an Alloc, so we need to
-			// identify the Source from the Extract.
-			case *ssa.Extract:
-				t := v.Tuple.Type().(*types.Tuple).At(v.Index).Type()
-				if _, ok := t.(*types.Pointer); ok && IsSourceType(conf, taggedFields, t) {
-					sources = append(sources, New(v))
-				}
-				continue
-
-			// value received from chan
-			case *ssa.UnOp:
-				// not a <-chan operation
-				if v.Op != token.ARROW {
-					continue
-				}
-
-			// value obtained through a field or an index operation
-			case *ssa.Field, *ssa.FieldAddr, *ssa.Index, *ssa.IndexAddr, *ssa.Lookup:
-
-			// chan or map value (arrays and slices are created using Allocs)
-			case *ssa.MakeMap, *ssa.MakeChan:
-			}
-
-			// all of the instructions that the switch lets through are values as per ssa/doc.go
-			if v := instr.(ssa.Value); IsSourceType(conf, taggedFields, v.Type()) {
-				sources = append(sources, New(v.(ssa.Node)))
+			if n := instr.(ssa.Node); isSourceNode(n, conf, propagators, taggedFields) {
+				sources = append(sources, New(n))
 			}
 		}
 	}
 	return sources
 }
 
+func isSourceNode(n ssa.Node, conf *config.Config, propagators fieldpropagator.ResultType, taggedFields fieldtags.ResultType) bool {
+	switch v := n.(type) {
```

Comment:
Save for `Extract`, all of these cases have an identical call: `IsSourceType(conf, taggedFields, n.(ssa.Value).Type())`. 

Consider extracting a variable before the switch, e.g. `nodeHasSourceType = IsSourceType(...)` to help readability.

Reply:
> Save for Extract and TypeAssert, all of these cases have an identical call: IsSourceType(conf, taggedFields, n.(ssa.Value).Type()).

Crucially, the `default` case is doing some lifting here.  Non-`Value` instructions do not have `Type()`.  That's an easy enough check to add... But even so restricted, `Range` instructions cause panics.  From [their docs](https://pkg.go.dev/golang.org/x/tools/go/ssa#Range).

> Type() returns an opaque and degenerate "rangeIter" type.

So to extract that call would be to add the preamble:

```golang
func isSourceNode(n ssa.Node, conf *config.Config, propagators fieldpropagator.ResultType, taggedFields fieldtags.ResultType) bool {
	_, isVal := n.(ssa.Value)
	_, isRange := n.(*ssa.Range)
	if !isVal || isRange {
		return false
	}

	isSrcTyp := IsSourceType(conf, taggedFields, n.(ssa.Value).Type())
	switch v := n.(type) {
		// ...
	}
}
```

At that point, I'm not sure which is more readability.  Thoughts?

Reply:
Ah, good point. On closer look, I think the repetition is fine. I think I got distracted by the call itself, which looks quite onerous.